File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public void backtracking(int index, String curr, List<String> result, String digits, HashMap<Character, List<Character>> numberToLetter){
        if(curr.length() == digits.length()){
            result.add(curr);
            return;
        }
        for (Character i:  numberToLetter.get(digits.charAt(index)))
            backtracking(index + 1, curr + i, result, digits, numberToLetter);

    }

        
    public List<String> letterCombinations(String digits) {
        HashMap<Character, List<Character>> numberToLetter = new HashMap<Character, List<Character>>();
        numberToLetter.put('2', Arrays.asList('a','b','c'));
        numberToLetter.put('3', Arrays.asList('d','e','f'));
        numberToLetter.put('4', Arrays.asList('g','h','i'));
        numberToLetter.put('5', Arrays.asList('j','k','l'));
        numberToLetter.put('6', Arrays.asList('m','n','o'));
        numberToLetter.put('7', Arrays.asList('p','q','r','s'));
        numberToLetter.put('8', Arrays.asList('t','u','v'));
        numberToLetter.put('9', Arrays.asList('w','x','y','z'));
        
        List<String> result = new ArrayList<String>();
        if (digits.length() > 0)
            backtracking(0,"", result, digits, numberToLetter);
        return result;
    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        numberToLetter = {
        '2':['a','b','c'],
        '3':['d','e','f'],
        '4':['g','h','i'],
        '5':['j','k','l'],
        '6':['m','n','o'],
        '7':['p','q','r','s'],
        '8':['t','u','v'],
        '9':['w','x','y','z']
        }
        result = []

        def backtracking(index, curr):
            if len(curr) == len(digits):
                result.append(curr)
                return
            for i in numberToLetter[digits[index]]:
                backtracking(index + 1, curr + i)
        
        if digits:
            backtracking(0,"")
        return result